transaction/src/transaction.rs

```diff
@@ -42,6 +46,49 @@ pub struct Transaction {
 }
 
 impl Transaction {
+    pub fn payload_keys(
+        &self,
+        fvk: &FullViewingKey,
+    ) -> anyhow::Result<BTreeMap<Commitment, PayloadKey>> {
+        let mut result = BTreeMap::new();
+
+        for action in self.actions() {
+            match action {
+                Action::Swap(swap) => {
+                    let epk = &swap.body.swap_nft.ephemeral_key;
+                    let shared_secret = fvk.incoming().key_agreement_with(epk)?;
+                    let payload_key = PayloadKey::derive(&shared_secret, epk);
+                    let commitment = swap.body.swap_nft.note_commitment;
+
+                    result.insert(commitment, payload_key);
+                }
+                Action::SwapClaim(swap_claim) => {
+                    let epk_1 = &swap_claim.body.output_1.ephemeral_key;
+                    let epk_2 = &swap_claim.body.output_2.ephemeral_key;
+                    let shared_secret_1 = fvk.incoming().key_agreement_with(epk_1)?;
+                    let shared_secret_2 = fvk.incoming().key_agreement_with(epk_2)?;
+                    let payload_key_1 = PayloadKey::derive(&shared_secret_1, epk_1);
+                    let payload_key_2 = PayloadKey::derive(&shared_secret_2, epk_2);
+                    let commitment_1 = swap_claim.body.output_1.note_commitment;
+                    let commitment_2 = swap_claim.body.output_2.note_commitment;
+
+                    result.insert(commitment_1, payload_key_1);
+                    result.insert(commitment_2, payload_key_2);
+                }
+                Action::Output(output) => {
+                    let epk = &output.body.note_payload.ephemeral_key;
+                    let shared_secret = fvk.incoming().key_agreement_with(epk)?;
+                    let payload_key = PayloadKey::derive(&shared_secret, epk);
+                    let commitment = output.body.note_payload.note_commitment;
+
+                    result.insert(commitment, payload_key);
+                }
+                _ => {}
+            }
+        }
```

Comment:
Are these the only actions that require payload keys? Will there be others in the future?

The wildcard match here makes me a little uncomfortable because I worry we'll forget to update this function. One way to ensure we don't forget to update it would be to delegate the extraction of payload keys to a separate function on each action type, and invoke each (many of them would not do anything). This would ensure that if we add new actions, we won't forget to update this function.